output/filewriter_test.go

```diff
@@ -62,3 +80,398 @@ func TestFileWriterConfig(t *testing.T) {
+	var records []string
+	for _, name := range decompressed {
+		f, err := os.Open(name)
+		if err != nil {
+			t.Fatal(err)
+		}
+		scan := bufio.NewScanner(f)
+		for scan.Scan() {
+			records = append(records, scan.Text())
+		}
+		f.Close()
+	}
+	sort.Strings(records)
+
+	var out []byte
+	for _, rec := range records {
+		out = append(out, rec...)
+		out = append(out, '\n')
+	}
+
+	testutil.DiffWithGolden(t, out, filepath.Join("testdata", "filewriter", "input.sorted.golden"))
+	if t.Failed() {
+		outName := filepath.Join(os.TempDir(), t.Name())
+		fmt.Printf("ERROR: writing incorrect buffer to %q for investigation\n\n", outName)
+		if err := os.WriteFile(outName, out, os.ModePerm); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func testFileWriterIntegration(t *testing.T, tmpDir, pathString string, procs int, rotate time.Duration) []string {
+	// This test uses a randomly generated input CSV file.
+	//  schema: pick(AAA|BBB|CCC|DDD), digit(22), first, last, email, state
+	//  site: https://www.convertcsv.com/generate-test-data.htm
+
+	toml := `
+	[fields]
+	names = ["kind", "digits", "first", "last", "email", "state"]
+
+	[input]
+	name = "list"
+	
+	[input.config]
+	files = ["./testdata/filewriter/input.csv.log.zst"]
+	
+	[output]
+	fields = ["kind"]
+	name = "filewriter"
+	procs = %d
+	[output.config]
+	pathstring = %q
+	rotateinterval = %q
+`
+	if !testing.Verbose() {
+		defer testutil.LessLogging()()
+	}
+
+	toml = fmt.Sprintf(toml, procs, strings.Replace(pathString, "TMPDIR", tmpDir, -1), rotate)
+
+	cfg, err := baker.NewConfigFromToml(strings.NewReader(toml),
+		baker.Components{
+			Inputs:  []baker.InputDesc{input.ListDesc},
+			Outputs: []baker.OutputDesc{output.FileWriterDesc},
+		})
+	if err != nil {
+		t.Fatal(err)
+	}
+	topo, err := baker.NewTopologyFromConfig(cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	topo.Start()
+	topo.Wait()
+
+	// We decompress files since the content of compressed files is not
+	// guaranteed to be determinsitic, however it's lossless so we'll use the
+	// decompressed file to control the files the output produced.
+	return decompressFilesInDir(t, tmpDir)
+}
+
+func TestFileWriterIntegrationField0(t *testing.T) {
+	testFileWriterIntegrationDeterministic(t, filepath.Join("TMPDIR", "{{.Field0}}", "out.csv.zst"))
+}
+
+func TestFileWriterIntegrationIndex(t *testing.T) {
+	testFileWriterIntegrationDeterministic(t, filepath.Join("TMPDIR", "{{.Index}}", "subdir", "out.csv.zst"))
+}
+
+func TestFileWriterIntegrationRotationIndex(t *testing.T) {
+	testFileWriterIntegrationDeterministic(t, filepath.Join("TMPDIR", "{{.Rotation}}", "out.csv.zst"))
+}
+
+func TestFileWriterIntegrationTimestamp(t *testing.T) {
+	const procs = 1
+	const rotate = -1
+	testFileWriterIntegrationCheckRecords(t, filepath.Join("TMPDIR", "{{.Year}}{{.Month}}{{.Day}}{{.Hour}}{{.Minute}}{{.Second}}-out.csv.zst"), procs, rotate)
+}
+func TestFileWriterIntegrationProcs(t *testing.T) {
```

Comment:
Here is missing a test in which both `index` and `field0` are used. This is the situation in which we have the greater number of concurrent workers `nproc` FileWriter times a fileWorker for each `field0` occurrence. Maybe is a case worth testing.

Reply:
👍🏽

Reply:
done